### PR TITLE
LPS-114977 As a developer I can specify custom tokens so that the style editor provides a form to edit those tokens instead of the Liferay-standard ones

### PR DESCRIPTION
**Pull request by izaera**
# Architecture overview

## Introduction

The feature is currently split in two levels: 

- Frontend infrastructure:
	1. Provides a `DynamicInclude` to inject CSS variables in the HTML
	2. Provides a registry that describes configurable CSS variables per deployed theme
	
- Layouts Administration: 
	1. Manages user configured variables for themes on a `LayoutSet` basis 
	2. Provides the variable values to the infra level so that they are injected in the HTML

### frontend-css-variables-api

- `ThemeCSSVariableDescriptionsRegistry`: a service that provides CSS variable descriptions for deployed themes. Variable descriptions are stored in `CSSVariableDescription` objects, which have a type field of class `CSSVariableType`.
- `ScopedCSSVariablesProvider`: a service interface that may be implemented by anyone wanting to inject CSS variable values in the HTML. The implementers receive the `HttpServletRequest` and return a `List` of `ScopedCSSVariable`s, which are objects holding the CSS scope where variables should be placed and their values.

### frontend-css-variables-web

- `ThemeCSSVariableDescriptionsRegistryImpl`: implements the `ThemeCSSVariableDescriptionsRegistry` service.
- `ScopedCSSVariablesTopHeadDynamicInclude`: asks `ScopedCSSVariablesProvider` instances for CSS variables and injects them in the HTML.

### frontend-theme-classic

- This is the default Classic theme, where we are adding a `css-variables.json` file to describe two configuration variables.

### layout-admin-api

- `LayoutSetCSSVariablesConfiguration`: a service to manage CSS variable definitions for a given `LayoutSet` (note that given the `LayoutSet` we know the `Theme` and the `Company`, so we can store different definitions for each `(LayoutSet, Theme, Company)` tuple. This service could be hidden, but I preferred to make it public for now in case we want to configure or obtain the CSS variables from other parts of the portal. However, in the final release, we should decide if we want this to be a public API or just a functionality only available through the GUI.

### layout-admin-web

- `LayoutScopedCSSVariablesProvider` and `LayoutScopedCSSVariables`: an implementation of `ScopedCSSVariablesProvider` and `ScopedCSSVariables` that returns CSS variables for the request's `LayoutSet` to the `ScopedCSSVariablesTopHeadDynamicInclude` service.
- `LayoutSetCSSVariablesConfigurationImpl`: implementation of `LayoutSetCSSVariablesConfiguration` that uses Company Settings as its configuration store.
- `GroupPagesPortlet` and `look_and_feel_theme_details.jsp`: access the `LayoutSetCSSVariablesConfiguration` and `ThemeCSSVariableDescriptionsRegistry` to show a form to edit the CSS variables for a given `LayoutSet`.
- `EditLayoutSetMVCActionCommand`: receives the values of the `look_and_feel_theme_details.jsp` form and stores them using the `LayoutSetCSSVariablesConfiguration` service.

### CODE REVIEW

**Comment by jbalsas** [comment]
Hey @izaera, this looks good at least as a POC!

Could you please draw a simplified schema of the infrastructure in place here? Where and how are we storing data, how are we presenting the edition UI and where and how are we then dumping the variables?

This last part I think we can still flesh out a bit more: https://github.com/jbalsas/liferay-portal/pull/2211/files#diff-e2a1b23a5b6a7fd84841fdf7dfa05e22R60

So, instead of relying on this particular configuration set... we should have some type of mechanism so others could register a set of values and a scope. 

That would allow, for instance, someone to register a set of `[x, y, z]` variables to the `#_portletId_` scope and we could print them all out at once. This should be useful when dealing with site creation (layouts, widgets, fragments, experiences).

**Comment by izaera** [comment]
> So, instead of relying on this particular configuration set... we should have some type of mechanism so others could register a set of values and a scope.

:thinking:  What you are asking for here is?

1) Let variables be instantiated in scopes other than organization (for example: let users assign a variable a value for a specific portlet that overrides the value for the company)
2) Let variables be declared in places other than a theme WAR (for example: let users deploy an OSGi bundle that declares new variables for a given theme-portlet pair, or other pairs/tuples we define)
3) Both 1 and 2
4) Other thing

**Comment by izaera** [comment]
I have pushed the new changes requested as per private conversation in Slack (see the two new commits)

File: modules/apps/frontend-css/frontend-css-variables-api/src/main/java/com/liferay/frontend/css/variables/theme/ThemeCSSVariableDescriptionsRegistry.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+import com.liferay.frontend.css.variables.CSSVariableDescription;
+import com.liferay.portal.kernel.model.Theme;
+
+import java.util.Map;
+
+/**
+ * @author Iván Zaera Avellón
+ */
+public interface ThemeCSSVariableDescriptionsRegistry {
```
Comment (jbalsas) [inline modules/apps/frontend-css/frontend-css-variables-api/src/main/java/com/liferay/frontend/css/variables/theme/ThemeCSSVariableDescriptionsRegistry.java:25]: What's the meaning of `Theme`, here? What's this interface for?

**Comment by jbalsas** [comment]
Hey @izaera, even though we've talked about this, it's hard for me to fully visualize what's what straight from the code.. could you UML this for me? 😂 

Just kidding, no need to UML it, but at least give me a diagram or explanation of the architecture?

Right now I'm mostly interested in the part that allows external apps to contribute variable values that we output... let's focus on that subsystem and try to send that independently?

**Comment by jbalsas** [comment]
I was also thinking about UI generation... I'm wondering if we can simply use the React version of formbuilder and be done with server side... 🤔

**Comment by izaera** [comment]
> Right now I'm mostly interested in the part that allows external apps to contribute variable values that we output... let's focus on that subsystem and try to send that independently?

Oops, forgot about that :-(

File: modules/apps/frontend-css/frontend-css-variables-api/src/main/java/com/liferay/frontend/css/variables/theme/ThemeCSSVariableDescriptionsRegistry.java (hunk 1 above)
Comment (izaera) [reply modules/apps/frontend-css/frontend-css-variables-api/src/main/java/com/liferay/frontend/css/variables/theme/ThemeCSSVariableDescriptionsRegistry.java:25]: I'll explain it in the architecture overview (below).

**Comment by izaera** [comment]
Here it is (the architecture explanation) :point_down: 

# Architecture overview

## Introduction

The feature is currently split in two levels: 

- Frontend infrastructure:
	1. Provides a DynamicInclude to inject CSS variables in the HTML
	2. Provides a registry that describes configurable CSS variables per deployed theme
	
- Layouts Administration: 
	1. Manages user configured variables for themes on a `LayoutSet` basis 
	2. Provides the variable values to the infra level so that they are injected in the HTML


## Projects 

### frontend-css-variables-api

- `ThemeCSSVariableDescriptionsRegistry`: a service that provides CSS variable descriptions for deployed themes. Variable descriptions are stored in a `CSSVariableDescription` object, which has a type field of class `CSSVariableType`.
- `ScopedCSSVariablesProvider`: a service interface that may be implemented by anyone wanting to inject CSS variable values in the HTML. The implementers receive the `HttpServletRequest` and return a `List` of `ScopedCSSVariable`s, which are objects holding the CSS scope where variables should be placed and their values.

### frontend-css-variables-web

- `ThemeCSSVariableDescriptionsRegistryImpl`: implements the `ThemeCSSVariableDescriptionsRegistry` service.
- `ScopedCSSVariablesTopHeadDynamicInclude`: asks `ScopedCSSVariablesProvider` instances for CSS variables and injects them in the HTML.

### frontend-theme-classic

- This is the default Classic theme, where we are adding a `css-variables.json` file to describe two configuration variables.

### layout-admin-api

- `LayoutSetCSSVariablesConfiguration`: a service to manage CSS variable definitions for a given `LayoutSet` (note that given the `LayoutSet` we know the `Theme` and the `Company`, so we can store different definitions for each `(LayoutSet, Theme, Company)` tuple. This service could be hidden, but I preferred to make it public for now in case we want to configure or obtain the CSS variables from other parts of the portal. However, in the final release, we should decide if we want this to be a public API or just a functionality only available through the GUI.

### layout-admin-web

- `LayoutScopedCSSVariablesProvider` and `LayoutScopedCSSVariables`: an implementation of `ScopedCSSVariablesProvider` and `ScopedCSSVariables` that returns CSS variables for the request's `LayoutSet` to the `ScopedCSSVariablesTopHeadDynamicInclude` service.
- `LayoutSetCSSVariablesConfigurationImpl`: implementation of `LayoutSetCSSVariablesConfiguration` that uses Company Settings as its configuration store.
- `GroupPagesPortlet` and `look_and_feel_theme_details.jsp`: access the `LayoutSetCSSVariablesConfiguration` and `ThemeCSSVariableDescriptionsRegistry` to show a form to edit the CSS variables for a given `LayoutSet`.
- `EditLayoutSetMVCActionCommand`: receives the values of the `look_and_feel_theme_details.jsp` form and stores them using the `LayoutSetCSSVariablesConfiguration` service.

**Comment by izaera** [comment]
I have rebased on top of master and force-pushed. You may need to redownload the PR.

**Comment by jbalsas** [comment]
Let's focus on `ScopedCSSVariablesProvider` as an initial thing we can land.

`ThemeCSSVariableDescriptionsRegistry` naming bothers me slightly... can't exactly say why...

**Comment by jbalsas** [comment]
Closing here, since we've been moving this around independently

**Pull request closed without merging**